rust/src/mqtt/detect.rs

```diff
@@ -17,421 +17,1528 @@
-    MQTT_DONT_CARE = 0,
-    MQTT_MUST_BE_SET = 1,
-    MQTT_CANT_BE_SET = 2,
-}
-
-#[inline]
-fn check_flag_state(flag_state: MQTTFlagState, flag_value: bool, ok: &mut bool) {
-    match flag_state {
-        MQTTFlagState::MQTT_MUST_BE_SET => {
-            if !flag_value {
-                *ok = false;
-            }
-        }
-        MQTTFlagState::MQTT_CANT_BE_SET => {
-            if flag_value {
-                *ok = false;
-            }
-        }
-        _ => {}
-    }
-}
-
-#[no_mangle]
-pub extern "C" fn rs_mqtt_tx_has_type(tx: &MQTTTransaction, mtype: u8) -> u8 {
-    for msg in tx.msg.iter() {
-        if mtype == msg.header.message_type as u8 {
-            return 1;
-        }
-    }
-    return 0;
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn rs_mqtt_cstr_message_code(
-    str: *const std::os::raw::c_char,
-) -> std::os::raw::c_int {
-    let msgtype: &CStr = CStr::from_ptr(str);
-    if let Ok(s) = msgtype.to_str() {
-        if let Ok(x) = MQTTTypeCode::from_str(s) {
-            return x as i32;
-        }
-    }
-    return -1;
-}
-
-#[no_mangle]
-pub extern "C" fn rs_mqtt_tx_has_flags(
-    tx: &MQTTTransaction, qretain: MQTTFlagState, qdup: MQTTFlagState,
-) -> u8 {
-    for msg in tx.msg.iter() {
-        let mut ok = true;
-        check_flag_state(qretain, msg.header.retain, &mut ok);
-        check_flag_state(qdup, msg.header.dup_flag, &mut ok);
-        if ok {
-            return 1;
-        }
-    }
-
-    return 0;
-}
-
-#[no_mangle]
-pub extern "C" fn rs_mqtt_tx_has_qos(tx: &MQTTTransaction, qos: u8) -> u8 {
+fn mqtt_tx_has_type(tx: &MQTTTransaction, mtype: &DetectUintData<u8>) -> c_int {
     for msg in tx.msg.iter() {
-        if qos == msg.header.qos_level {
+        if detect_match_uint(mtype, msg.header.message_type as u8) {
             return 1;
         }
     }
     return 0;
 }
 
-#[no_mangle]
-pub extern "C" fn rs_mqtt_tx_get_protocol_version(state: &MQTTState) -> u8 {
-    return state.protocol_version;
-}
-
-#[no_mangle]
-pub extern "C" fn rs_mqtt_tx_has_connect_flags(
-    tx: &MQTTTransaction, username: MQTTFlagState, password: MQTTFlagState, will: MQTTFlagState,
-    will_retain: MQTTFlagState, clean_session: MQTTFlagState,
-) -> u8 {
-    for msg in tx.msg.iter() {
-        if let MQTTOperation::CONNECT(ref cv) = msg.op {
-            let mut ok = true;
-            check_flag_state(username, cv.username_flag, &mut ok);
-            check_flag_state(password, cv.password_flag, &mut ok);
-            check_flag_state(will, cv.will_flag, &mut ok);
-            check_flag_state(will_retain, cv.will_retain, &mut ok);
-            check_flag_state(clean_session, cv.clean_session, &mut ok);
-            if ok {
-                return 1;
-            }
-        }
-    }
-    return 0;
-}
-
-#[no_mangle]
-pub unsafe extern "C" fn rs_mqtt_tx_get_connect_clientid(
-    tx: &MQTTTransaction, buffer: *mut *const u8, buffer_len: *mut u32,
-) -> u8 {
+unsafe extern "C" fn mqtt_tx_get_connect_clientid(
```

Comment:
style: function naming

Reply:
@jasonish how do I prevent this function from being exported in headers ?

Reply:
Sorry, they are not exported.

Victor, how should I name these internal rust functions ?

Reply:
cf 
```
warning: function `SCMqttGetConnectClientId` should have a snake case name
  --> src/mqtt/detect.rs:56:22
   |
56 | unsafe extern "C" fn SCMqttGetConnectClientId(
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^ help: convert the identifier to snake case: `scmqtt_get_connect_client_id`
   |
   = note: `#[warn(non_snake_case)]` on by default
```

Reply:
Why are they `unsafe extern "C"` if not exported?

Reply:
Because they are callbacks, called from C

Reply:
That makes them exported, right?

Reply:
No, if only used via function pointer from the same file, they are not exported. No need to follow C naming conventions as they do not live in that name space. `pub extern` is when you enter the C name space.

Reply:
I see, thanks for the explanation. So for these we can use the rust style then, agree?

Reply:
> I see, thanks for the explanation. So for these we can use the rust style then, agree?

Yes, and so does clippy.. If pub extern it doesn't complain about style, but if just extern C it will.  It also means you can avoid the stutter, name it like a pure rust function instead of long prefixes if you like.